gadget-container/gadgets/collector-process/bpf/collector-process.c

```diff
@@ -0,0 +1,112 @@
+// SPDX-License-Identifier: GPL-2.0
+/* Copyright (c) 2021 The Inspektor Gadget authors */
+
+/* Inspired by the BPF iterator in the Linux tree:
+ * https://github.com/torvalds/linux/blob/v5.12/tools/testing/selftests/bpf/progs/bpf_iter_task.c
+ */
+
+/* This BPF program uses the GPL-restricted function bpf_seq_printf().
+ */
+
+#include <linux/bpf.h>
+
+#include <bpf/bpf_helpers.h>
+
+/* libbpf v0.4.0 introduced BPF_SEQ_PRINTF in bpf_tracing.h.
+ * In future versions, it will be in bpf_helpers.h.
+ */
+#include <bpf/bpf_tracing.h>
+
+#include "../../../../pkg/gadgettracermanager/gadget-tracer-manager-maps.h"
+
+struct bpf_iter_meta {
+	__bpf_md_ptr(struct seq_file *, seq);
+	__u64 session_id;
+	__u64 seq_num;
+};
+
+struct bpf_iter__task {
+	struct bpf_iter_meta *meta;
+	struct task_struct *task;
+} __attribute__((preserve_access_index));
+
+struct nsproxy {
+	struct mnt_namespace *mnt_ns;
+} __attribute__((preserve_access_index));
+
+struct task_struct {
+	int pid;
+	int tgid;
+	char comm[16];
+	struct nsproxy *nsproxy;
+} __attribute__((preserve_access_index));
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, __u64);
+	__type(value, __u64);
+	__uint(max_entries, 128);
+} context SEC(".maps");
+
+SEC("iter/task")
+int dump_task(struct bpf_iter__task *ctx)
+{
+	struct seq_file *seq = ctx->meta->seq;
+	__u32 seq_num = ctx->meta->seq_num;
+	__u64 session_id = ctx->meta->session_id;
+	struct task_struct *task = ctx->task;
+	__u64 *counter;
+	__u64 zero = 0;
+
+	if (seq_num == 0) {
+		if (task == (void *)0) {
+			/* Empty list */
+			BPF_SEQ_PRINTF(seq, "[\n]\n");
+			return 0;
+		} else {
+			BPF_SEQ_PRINTF(seq, "[\n");
+		}
+		bpf_map_update_elem(&context, &session_id, &zero, BPF_ANY);
+	}
+
+	counter = bpf_map_lookup_elem(&context, &session_id);
+	if (!counter)
+		return 0;
+
+	if (task == (void *)0) {
+		if (*counter)
+			BPF_SEQ_PRINTF(seq, "\n");
+		BPF_SEQ_PRINTF(seq, "]\n");
+		bpf_map_delete_elem(&context, &session_id);
+		return 0;
+	}
+
+        __u64 mntns_id = task->nsproxy->mnt_ns->ns.inum;
+
+#ifdef WITH_FILTER
+	__u32 *found = bpf_map_lookup_elem(&filter, &mntns_id);
+	if (!found)
+		return 0;
+#endif
+
+	if (*counter)
+		BPF_SEQ_PRINTF(seq, ",\n");
+
+	__sync_fetch_and_add(counter, 1);
```

Comment:
This is because some processes can be filtered out, so `seq_num==0` is not a good test to know if this is the first entry to be printed because maybe the first process was filtered out, and then we must not print the `,` before the second process; that would result in invalid json.